### ISSUE TITLE
[JIT] TensorExpr causes segmentation fault

### ISSUE THREAD

**Issue by xwang233** [pytorch/pytorch#46440]
## 🐛 Bug

JIT TensorExpr causes segmentation fault. Bisection shows it's introduced in https://github.com/pytorch/pytorch/pull/45396. 

original discussion link https://github.com/pytorch/pytorch/issues/43429#issuecomment-709665059

## To Reproduce

- git clone nvidia/deeplearningexample
- checkout `ac059022a54051df1ca4a9c55e0d27c475a6999d`
- cd `PyTorch/SpeechSynthesis/Tacotron2`
- run `python inference_perf.py -m Tacotron2 -bs 1 --decoder-no-early-stopping --synth-data true -o .`

The core dump file says the segfault is at 
```cpp
torch::jit::tensorexpr::ExprHandle::operator-(torch::jit::tensorexpr::ExprHandle const&) const ()
```

The full stack trace is at https://gist.githubusercontent.com/xwang233/120552174810acaa92411826db8f32c1/raw/ccbf397a05cece5dd4fe6794a75eea07f39249d4/stack-trace-43429.txt

## Expected behavior

no crash

## Environment

```
Collecting environment information...
PyTorch version: 1.8.0a0+fc1d6bf
Is debug build: True
CUDA used to build PyTorch: 11.0
ROCM used to build PyTorch: N/A

OS: Fedora 29 (Workstation Edition) (x86_64)
GCC version: (GCC) 8.3.1 20190223 (Red Hat 8.3.1-2)
Clang version: 7.0.1 (Fedora 7.0.1-6.fc29)
CMake version: version 3.14.5

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: TITAN V
GPU 1: TITAN V

Nvidia driver version: 450.36
cuDNN version: Probably one of the following:
/usr/lib64/libcudnn.so.8.0.0
/usr/lib64/libcudnn_adv_infer.so.8.0.0
/usr/lib64/libcudnn_adv_train.so.8.0.0
/usr/lib64/libcudnn_cnn_infer.so.8.0.0
/usr/lib64/libcudnn_cnn_train.so.8.0.0
/usr/lib64/libcudnn_ops_infer.so.8.0.0
/usr/lib64/libcudnn_ops_train.so.8.0.0
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.8.0a0
[pip3] torchvision==0.8.0a0+6e639d3
[conda] Could not collect
```

## Additional context

added to 1.7.0 milestone as requested by @ngimel

cc @gmagogsfm @ptrblck @ngimel

**Comment by ngimel** [comment]
cc @ZolotukhinM, @Krovatkin

**Comment by ZolotukhinM** [comment]
Thanks for the report, I'll take a look at it!

I've extracted a smaller repro if anyone is curious:
```
import torch

ir = """
graph(%0 : Float(1, 256, strides=[256, 1], requires_grad=0, device=cuda:0),
      %1 : Float(1, 512, strides=[512, 1], requires_grad=0, device=cuda:0),
      %2 : Float(1, 140, strides=[140, 1], requires_grad=0, device=cuda:0),
      %3 : Float(1, 140, strides=[140, 1], requires_grad=0, device=cuda:0)):
  %4 : int = prim::Constant[value=1]()
  %5 : int = prim::Constant[value=1]()
  %6 : Float(1, 1, 140, strides=[140, 140, 1], requires_grad=0, device=cuda:0) = aten::unsqueeze(%3, %5)
  %7 : int = prim::Constant[value=1]()
  %8 : int = prim::Constant[value=1]()
  %9 : Float(1, 1, 140, strides=[140, 140, 1], requires_grad=0, device=cuda:0) = aten::unsqueeze(%2, %8)
  %10 : Tensor[] = prim::ListConstruct(%6, %9)
  %11 : int = prim::Constant[value=1]()
  %12 : int = prim::Constant[value=1]()
  %attention_weights_cat.2 : Float(1, 2, 140, strides=[280, 140, 1], requires_grad=0, device=cuda:0) = aten::cat(%10, %12)
  %14 : Tensor[] = prim::ListConstruct(%0, %1)
  %15 : int = prim::Constant[value=-1]()
  %cell_input.2 : Float(1, 768, strides=[768, 1], requires_grad=0, device=cuda:0) = aten::cat(%14, %15)
  return (%cell_input.2, %attention_weights_cat.2)
"""
g = torch._C.parse_ir(ir)

torch._C._jit_pass_fuse_tensorexprs(g)
print(g)
inputs = [torch.randn(1, 256, device='cuda'),
          torch.randn(1, 512, device='cuda'),
          torch.randn(1, 140, device='cuda'),
          torch.randn(1, 140, device='cuda')]
x = torch._C._jit_interpret_graph(g, tuple(inputs))
```

**Comment by ZolotukhinM** [comment]
The issue seems to be with how we handle negative values of `dim` for `torch.cat` when converting it to tensor expressions. The code assumed that it cant be negative, but in this test it is. I'll submit a fix for this shortly.
